pkg/controller/controller_instance.go

```diff
@@ -370,81 +389,133 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
 	}
-	resp, rc, err := brokerClient.PollServiceInstance(instance.Spec.ExternalID, lastOperationRequest)
+
+	glog.V(5).Infof("Polling last operation on Instance %v/%v", instance.Namespace, instance.Name)
+
+	response, err := brokerClient.PollLastOperation(request)
 	if err != nil {
-		glog.Warningf("Poll failed for %v/%v  : %s", instance.Namespace, instance.Name, err)
-		return err
-	}
-	glog.V(4).Infof("Poll for %v/%v returned %q : %q", instance.Namespace, instance.Name, resp.State, resp.Description)
+		// If the operation was for delete and we receive a http.StatusGone,
+		// this is considered a success as per the spec, so mark as deleted
+		// and remove any finalizers.
+		if osb.IsGoneError(err) && deleting {
+			clone, err := api.Scheme.DeepCopy(instance)
+			if err != nil {
+				return err
+			}
+			toUpdate := clone.(*v1alpha1.Instance)
 
-	// If the operation was for delete and we receive a http.StatusGone,
-	// this is considered a success as per the spec, so mark as deleted
-	// and remove any finalizers.
-	if rc == http.StatusGone && deleting {
-		instance.Status.AsyncOpInProgress = false
-		// Clear the finalizer
-		if finalizers := sets.NewString(instance.Finalizers...); finalizers.Has(v1alpha1.FinalizerServiceCatalog) {
-			finalizers.Delete(v1alpha1.FinalizerServiceCatalog)
-			c.updateInstanceFinalizers(instance, finalizers.List())
+			toUpdate.Status.AsyncOpInProgress = false
+			c.updateInstanceCondition(
+				toUpdate,
+				v1alpha1.InstanceConditionReady,
+				v1alpha1.ConditionFalse,
+				successDeprovisionReason,
+				successDeprovisionMessage,
+			)
+
+			// Clear the finalizer
+			if finalizers := sets.NewString(toUpdate.Finalizers...); finalizers.Has(v1alpha1.FinalizerServiceCatalog) {
+				finalizers.Delete(v1alpha1.FinalizerServiceCatalog)
+				c.updateInstanceFinalizers(toUpdate, finalizers.List())
+			}
+
+			c.recorder.Event(instance, api.EventTypeNormal, successDeprovisionReason, successDeprovisionMessage)
+			glog.V(5).Infof("Successfully deprovisioned Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
+			return nil
 		}
-		c.updateInstanceCondition(
-			instance,
-			v1alpha1.InstanceConditionReady,
-			v1alpha1.ConditionFalse,
-			successDeprovisionReason,
-			successDeprovisionMessage,
-		)
-		c.recorder.Event(instance, api.EventTypeNormal, successDeprovisionReason, successDeprovisionMessage)
-		glog.V(5).Infof("Successfully deprovisioned Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
-		return nil
+
+		// We got some kind of error from the broker.  While polling last
+		// operation, this represents an invalid response and we should
+		// continue polling last operation.
+		//
+		// The ready condition on the instance should already have
+		// condition false; it should be sufficient to create an event for
+		// the instance.
+		errText := ""
+		if osb.IsHTTPError(err) {
```

Comment:
Some sense made. It's less extract method and more be really explicit about state transitions. If I come in here, I basically need to look for "state is my current state" before I can start parsing. but i've got to mentally go through and discard 85% of the code by reading if this if that. not even if-else.

I'd almost prefer if it was if-else and resubmit to queue, because then I could find the "one clause" that I need that could possibly execute.

I'm not advocating single exit policy, but it's very complicated here, and we're really not doing anything very complex.